src/vs/workbench/contrib/mcp/browser/mcpResourceQuickAccess.ts

```diff
@@ -391,21 +395,45 @@ export abstract class AbstractMcpResourceAccessPick {
 		}));
 
 		store.add(picker.onDidAccept(async event => {
-			if (!event.inBackground) {
-				picker.hide(); // hide picker unless we accept in background
-			}
-
-			if (runOptions?.handleAccept) {
-				runOptions.handleAccept?.(picker.activeItems[0], event.inBackground);
-			} else {
-				const [item] = picker.selectedItems;
-				const uri = await helper.toURI((item as ResourceQuickPickItem).resource);
-				if (uri) {
-					this._editorService.openEditor({ resource: uri, options: { preserveFocus: event.inBackground } });
+			try {
+				picker.busy = true;
+				if (runOptions?.handleAccept) {
+					runOptions.handleAccept?.(picker.activeItems[0], event.inBackground);
+				} else {
+					const [item] = picker.selectedItems;
+					let uri: URI | undefined;
+					if (isDirectoryMode) {
+						uri = URI.parse((item as IQuickPickItem).id!);
+					} else {
+						uri = await helper.toURI((item as ResourceQuickPickItem).resource);
+					}
+					// If the URI is a directory, we need to read its contents and show them in the picker.
+					// directory URIs end with a `/`
+					if (uri?.path.endsWith('/')) {
```

Comment:
I think it's fine to do this unconditionally. The resource is going to get resolved as soon as you pick it anyway so we may as well `resolve` here all the time and then we don't need to demand any specific URI pattern from the server.